Blog Post: TypeScript and GraphQL union (#715)
the-guild-org/the-guild-website: pages/blog/typescript-types.mdx

import withArticle from '../../ui/blog/article'
export const meta = {

>   title: 'Back to front: TypeScript union types and how to use GraphQL Codegen to create type-safe',

Reviewer: Hmm I'm not sure I understand the title

Author: Charly suggested a new title:

`Getting the best of TypeScript and GraphQL: union types`

  author: 'tuvalSimha',
  tags: ['graphql', 'typescript', 'codegen'],
  date: '2022-07-20',
  description: 'TypeScript union types in use GraphQL schema, and how to use GraphQL Codegen to create type-safe.',

>   image: '/blog-assets/graphql-code-generator/cover.png'

Reviewer: Let's use a cover image with the article's title in it to make it more visible on the website and when shared on social media

}
export default withArticle(meta)
 
# Introduction

> The combination between TypeScript and GraphQL is great. Both TypeScript and GraphQL support the concept of discriminated unions.

Reviewer 2: ```suggestion
The combination between TypeScript and GraphQL can be very powerful. Both TypeScript and GraphQL support the concept of discriminated unions.
```

> The combination between TypeScript and GraphQL can be very powerful. Both TypeScript and GraphQL support the concept of discriminated unions.

Reviewer 3: ```suggestion
The combination of TypeScript and GraphQL can be very powerful. Both TypeScript and GraphQL support the concept of discriminated unions.
```

> In this short blog post, I want to share my experience in my last project and share one of the most powerful methods to create type-safe on code-time using TypeScript and GraphQL.

Reviewer: ```suggestion
In this short blog post, I want to share my last project experience and share one of the most powerful methods to bring type safety to your code using TypeScript and GraphQL.
```

Maybe something like this?

One of my assumptions in this blog post is that you know some TypeScript. If you don't, you can check out ["Basic Types"](https://www.typescriptlang.org/docs/handbook/basic-types.html) handbook.

> Surprisingly, if you have some experience in TypeScript or not, you can understand the Types concept very fast. So, let's talk a little bit about TypeScript union types, one of the most remarkable features of this language.

Reviewer 2: I'm not sure I understand the initial part of the sentence.
Are you surprised that whether you have the experience or not, you can understand the Types concept fast?

# What is a union type in TypeScript?

So TypeScript language is based on a type system.
TypeScript union operator can be used to define a variable that can hold multiple data types: integer or number, character, string, float, etc.

Here you can find a basic example:

```ts
let pizzaLover: string | boolean // pizzaLover can be a string or boolean type

pizzaLover = 'YES' // string type
pizzaLover = true // boolean type
```

[Here's a live example in TS Playground](https://www.typescriptlang.org/play?#code/DYUwLgBADglgXnAhgGQPYDcQCcBcEDOYWMAdgOYQA+EARqqqIiQLABQbsCKG2EAvBADkATQCiAZUEQA9NIJFSFMAE8oIDvCRpMWfhCIBXEDLl0GIJvtXrWsiAD0A-GyA)

# How to use union types in TypeScript

TypeScript can also be used to combine complex types such as interfaces.

> You can think about one type that contains some others types. Here simple example that combined 2 types: `Sushi` and `Pizza`, both of them can be united into one type by the name `Food`.

Reviewer 2: ```suggestion
You can think about one type that contains some others types. Here is a simple example that combines 2 types: `Sushi` and `Pizza`, both of them can be united into one type by the name `Food`.
```

```ts
type Sushi = {
  kind: 'Sushi'
  fish: string
  amount: number
}

type Pizza = {
  kind: 'Pizza'
  topping: string
  amount: number
}

type Food = Sushi | Pizza
```

[Here's a live example in TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAygrgZwBYEsoF4oG8CwAoKKAaxQDsATALigHJ5kUb9CAzFZahYAJzIHNmUAIYBbAPZxSwaqTgiARhG74AvvnyhIUAAooAXnqEZsgkhWo1dBoUwJRgYsGH6ce-QaIlSZcxcrxqeOp4mtAAYmJi5Mb0qFAAPjr6hsFAA)

> More Union type example with `string` and `string`:

Reviewer 2: ```suggestion
And another Union type example, this time with `string` and `string`:
```

```ts
type PizzaAddons = 'olives' | 'mushrooms' | 'pepperoni'

const myPreference: PizzaAddons = 'olives' // ok
const myPreference: PizzaAddons = 'pineapple' // error
```

[Here's a live example in TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBACglgLwQQwIIBN0HsB2BnKAXigHIsAbOANwjxKgB9SBbAVzwAsAnLLZuxqUhhIPHHBIBYAFAyAxrjzAozEDC4QAZhA045EAFyxEKDNnxFSFarXoB6O1CwBreYuWr1WnRD2HjSGiYipYkYHA4EMgi5BD2jjo8XDJAA)

# GraphQL union and TypeScript union type

> GraphQL is a query language that is used to define the API of a server. Like TypeScript or even Go language, GraphQL supports union and interface types. The type system in its core principles, so we can use it to make type-safe API calls. The goal is to use it to propagate our backend types to the frontend.

Reviewer 2: ```suggestion
GraphQL is a query language that is used to define the API of a server. Like TypeScript or even Go language, GraphQL supports union and interface types. The type system is one of its core principles and we can use it to make type-safe API calls. The goal is to use it to propagate our backend types to the frontend.
```

> When I just start to learn TypeScript, I found myself did a lot of types errors and annoying misspellings. Those error amounts became bigger and bigger when the project grow.

Reviewer 2: ```suggestion
When I started to learn TypeScript, I found myself making lots of type errors and annoying misspellings. These error became more frequent as the project grew.
```

> The best solution for this situation is to add a type-safety base GraphQL schema.

Reviewer 2: ```suggestion
The best solution for that situation was to add type-safety based on a GraphQL schema.
```

> GraphQL schema can return one of the multiple object types when we create Union or Interface. Like TypeScript, we can declare which object types are included in the union:

Reviewer 2: ```suggestion
A GraphQL schema can return one of multiple object types thanks to the Union or Interface keywords. Like TypeScript, we can declare which object types are included in the union:
```

```graphql
union Menu = Pizza | Sushi
```

Let's see a basic example schema that defines a `Menu` union type that can return either a `Pizza` or `Sushi`:

```graphql
union Menu = Pizza | Sushi

type Pizza {
  title: String!
  topping: String
}

type Sushi {
  name: String!
  fish: String
}

type Query {
  menu: [Menu!]
}
```

> After we declare our types or create a union type to return some of the others' types, we can generate our types from the schema and use them in the frontend.

Reviewer 2: ```suggestion
After we declare our types in GraphQL and create a union type to return some of the other types, we can generate TypeScript types from the GraphQL schema and use them in the frontend.
```

> After we declare our types or create a union type to return some of the others' types, we can generate our types from the schema and use them in the front.

Reviewer: ```suggestion
After we declare our types or create a union type to return some of the others' types, we can generate our types from the schema and use them in the frontend.
```

# How to use GraphQL Codegen to generate code and deal with TypeScript union types

First of all, you are more than welcome to explore all the plugins of GraphQL codegen. [GraphQL Code-Generator - LINK](https://graphql-code-generator.com/)
Like we did in the previous example, let's talk about getting the data typed safely in the front.

> (I use the `react-query` example)

Reviewer 2: ```suggestion
(in this case, I'll use the `react-query` example)
```

> Like we did in the previous example, let's talk about how to get the data typed safe in the front.

Reviewer: ```suggestion
Like we did in the previous example, let's talk about how to get the data typed safe in the frontend.
```

Create our query to get the `Menu`

```ts
const menuQuery = gql`
  query GetFullMenu {
    menu {
      ...Sushi
      ...Pizza
    }
  }
`
```

> Get the `Menu` type from the schema with the query, so we will be able to get the objects inside the `Menu`.

Reviewer 3: ```suggestion
After running codegen using the [`@graphql-codegen/typescript-graphql-request`](https://npmjs.com/package/@graphql-codegen/typescript-graphql-request) plugin, we end up having our result variable (`menu`) properly typed with an union type:
```

```ts
const Menu = () => {
  const { data } = useQuery<MenuQuery>('menu', async () => {
    const { menu } = await request(endpoint, menuQuery)
    return menu
  })
}
```

## Conclusion

> - I hope this blog post helps you better the why TypeScript unions and GraphQL unions and the way to use them in your project.

Reviewer 2: ```suggestion
- I hope this blog post helped you to better understand TypeScript unions and GraphQL unions and the way to use them in your project.
```

> - We can create Union type and Interface in TypeScript and GraphQL that help us to define the API of a server and use it in our client.

Reviewer 2: ```suggestion
- We can create Union and Interface types in both TypeScript and GraphQL, that helps us to better define the API of a server and use it in our client.
```

> - A static type check can help safeguard your application. Especially when you building large-scale applications.

Reviewer 2: ```suggestion
- A static type check can help safeguard your application. Especially in large-scale applications.
```

- When you use type-safe GraphQL, it helps you to eliminate existing errors like misspellings and errors, especially in code-time.

> - In this brief blog post, we looked at a few features of TypeScript that help you on code-time, most help you to understand the output of any function you create.

Reviewer 2: can we have a better name for `code-time`? not sure what...

It can be implemented using the `typeof` operator like we learn in the previous example. We can also use type guard with `in` operator to check if the variable is of the expected type.

```ts
interface Sushi {
  fish: number
}
interface Pizza {
  topping: string
}

function identifyFood(food: Sushi | Pizza) {
  if ('fish' in food) {
    food
    // ^?
  } else {
    food
    // ^?
  }
}
```

[Here's a live example in TS Playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMoFcDOALYyDeAsAFDLIzA4BcyI6AtgEbQDcJAviaJLIigArAAXkLgESZMAHsADjNABzGpjBRFbYp2IkY6EAjDApIZMAAmEcMBgBPAGJSpZgBQxHZmhhx4APskEicACU4qSmMMjOAOQUOFGmJm5OIURhZElmGmRkAPQ5yAB6APwSyOzIEAA2mCip2eTuWdl5hSVhWuwsQA)

## Conclusion

- I hope this blog post help you better the why of TypeScript types system work and how to use it for your projects.

- In this brief blog post, we looked at a few feature of TypeScript that help you on code-time, mosthly help you to understand the output of any function you create.

- All examples above allows us to build more versatile types by relying on a template pattern for the discriminant property rather than an exact string.

> - Recently i beed use GraphQL Codegen plugin to generate my types from my GraphQL schema. When my project is small, as a developer i can create and decalre any type that i want. But when the project is big, it hard to manage the types becuse we have too much types to create.

Reviewer 4: been?

  So the best way to manage the types, in my openion, is to use GraphQL Codegen plugin.

> ## So how do we creating Types based GraphQL schema?

Reviewer 4: I think this needs more background:

* GraphQL also has `union` and `interface` -> explain how it connected to the TS feature
* How GraphQL Codegen can help you -> add example

import withArticle from '../../ui/blog/article'
export const meta = {
  title: 'TypeScript basic types system',
  author: 'tuvalSimha',
  tags: ['graphql', 'typescript', 'codegen'],
  date: '2022-07-16',
  description: 'How to use Typescript basic types system',
  image: '/blog-assets/graphql-code-generator/cover.png'
}
export default withArticle(meta)
 
 
# Introduction

> TypeScript programming language is a superset of JavaScript that adds types to JavaScript using a set of tools called a type system.

Reviewer 4: I think the introduction is too broad - we don't have to talk about TypeScript or what it is, we can assume people who landed on this page know what TypeScript is. 

I would prefer the introduction to cover what's union types and then move to the actual thing.

All types inference assumes the expected type of the variable on code time. Type inference will log a complaint if the variable is later reassigned to a different type.
For Example:

```ts
let pizzaName = 'Napolitana'

pizzaName = 123 // Error: Type number (123) is not assignable to type 'string'.
```

When a variable is declared without being assigned an initial value, the default type is `any`.

```ts
let pizzaName

pizzaName = 123 // Any got all types of values
pizzaName = 'Napolitana' // Any got all types of values
```

Adding a type annpotation to a variable will force the variable to be of that type. It help you make sure the output will be the expected. And if not, you will get types error in code-time.

```ts
let pizzaName: string
pizzaName = 'Napolitana'

pizzaName = 123 // Error: Type number (123) is not assignable to type 'string'.
```

## Union Types

After we understand how TypeScript type system works, we can start to understand how to use it to create a Union Type.

> TypeScript allow us a flexible type that called `any` type like we did in the previous example. But also TypeScript can combine multiple types and combine them together as a union type.

Reviewer 4: Not sure I understand the note about `any`... Also I hate `any` so I don't think it should be mentioned as something you can/should use.

```ts
let pizzaLover: string | boolean

pizzaLover = 'YES' // string type
pizzaLover = true // boolean type
```

> More Union type example:

Reviewer 4: I thikn the jump from `string | boolean` to `A | B` is fine, but you should also mention here that TS supports string unions as well and give an example. 

```ts
type PizzaAddons = "olives" | "mushrooms" | "pepperoni";

const myPreference: PizzaAddons = "olives"; // ok
const myPreference: PizzaAddons = "pineapple"; // error
```

```ts
type Sushi = {
  kind: 'Sushi'
  fish: string
  amount: number
}

type Pizza = {
  kind: 'Pizza'
  topping: string
  amount: number
}

type Food = Sushi | Pizza
```

> So what is Type Narrowing?

Reviewer 4: Should be `type-narrowing`

Type Narrowing is a process of narrowing a type to a more specific type.
We can declare interface with all the properties of a type and then we can use that interface to narrow the type of a variable.
Lets make basic narrowing example with function:

```ts
interface Food {
  name: string
  nutritionalValues?: {
    calories: number
    fat: number
    protein: number
    carbs: number
  }
}

function getTotalCalories(f: Food): number | undefined {
  if (f.nutritionalValues) {
    f.nutritionalValues
    // f.nutritionalValues is type of { calories: number, fat: number, protein: number, carbs: number }
    return f.nutritionalValues.calories
  }
  f.nutritionalValues

  return f.nutritionalValues
}
```

### Operator `typeof`

Varibale of a union type can assume one or even few diffrent types. In the example below we use `if` statement to check if the variable is of type `Food` or not.
We can also use the opretor `typeof` with the variable name and compare it with the type you expect for the variable.
Lets see a basic example of `typeof` operator in function:

```ts

> function getScoreofLovingPizza(score: string | number) {

Reviewer 4: I think maybe we can try to provide a more meaningful example here. 
Something like `number | "very-much"` and then translate the "very-much" string into `10` as number?

  if (typeof score === 'string') {
    return console.log(`How much the use love pizza: ${score}`)
  } else {
    return console.log(`How much the use love pizza - Number score: ${score}`)
  }
}
```

### TypeScript Type Guard

Type Guard is a conditional statement that evaluates the type of a variable and if it is of the expected type, it will execute the code block.
It can be implemented using the `typeof` operator like we learn in the previous example. We can also use type guard with `in` operator to check if the variable is of the expected type.

```ts
type Sushi = {
  kind: 'Sushi'
  fish: string
  amount: number
}

type Pizza = {
  kind: 'Pizza'
  topping: string
  amount: number
}

type Food = Sushi | Pizza

function getFoodKindWithFish(Food: Fish | Bird) {
  if ('fish' in Food) {

>     return Food.fish()

Reviewer 4: Not sure I understand this? seems like a syntax error?
